### DISCUSSION TITLE
Consumer and Provider services specification examples data should be aligned in specific scenarios

### DISCUSSION THREAD

**Discussion by gurubabujampala** [specmatic/specmatic#1106] [Ask For Help] [answered]
In our Specmatic evaluation process, we've developed a sample use case featuring Customer Service as the requester and Order Service as the supplier.

**Use Case :** Customer Service initiates a request to the Order Service, seeking an order associated with a specific customer ID. Upon receiving the response, it verifies the response customer ID with input customer ID as part of validate the information.

**Business Logic:** 

1. Customer service is connecting to the order service and get the order details.
        **Order API**:  GET - /orders/customer/1001 (customerId)

       Got the below response from Order service based on order open api specification examples.
       
```javascript
[
    {
        "orderId": "ORD004",
        "customerId": 1004,
        "placeDate": '2022-01-01',
        "completionDate": '2022-01-02'
    },
    {
        "orderId": "ORD009",
        "customerId": 1082,
        "placeDate": '2024-04-22',
        "completionDate": '2024-04-23'
    }
]
``` 

2. Now, Customer service has a logic to check all the responses having input customerId.
In the above use case given customerId is 1001 and it does not match with any customerId in the response and it's fail.

Currently, we are aligning the examples across the customer and order open API specifications to ensure the success of this particular use case. However, achieving alignment in all instances poses a significant challenge. Is there a strategy we can employ to navigate through this situation effectively?

**Comment by harikrishnan83** [comment]
Hello @gurubabujampala , thanks for the details.

The inline examples in OpenAPI spec act as static stubs. Please have a look at [Specmatic dynamic expectations](https://specmatic.in/documentation/service_virtualization_tutorial.html#dynamic-expectations-aka-dynamic-stubbing-or-mocking---setting-expecations-over-specmatic-http-api). [Dynamic expectations can be set programmatically](https://specmatic.in/documentation/service_virtualization_tutorial.html#programmatically-starting-stub-server-within-tests) within your test (in `setup()` in case of contract tests and any relevant test method itself in case of component tests).

We are also working on a feature which will allow you to setup dynamic stubbing behaviour within the OpenAPI spec itself. We will share more on this shortly.

Hope this helps. Thanks.

**Reply by yogeshnikam671** [reply, marked as answer]
Hi @gurubabujampala, Specmatic now has the ability to substitute values in the response based on what is received in the request.
This is exactly what you are looking for in terms of `customerId`.
[Here](https://specmatic.io/documentation/service_virtualization_tutorial.html#inject-response-values-from-an-external-dictionary) is the documentation for the same.
Please take a look and let us know your feedback.
